### ISSUE TITLE
Add optional filtering to one visit per patient within assigned sets

### ISSUE THREAD

**Issue by zmek** [UCL-CORU/patientflow#100]
### What needs to be done?

# Add optional filtering to one visit per patient within assigned sets

## Summary
Add functionality to optionally filter data to one visit per patient within their assigned train/validation/test set, complementing the existing `single_snapshot_per_visit` functionality.

## Background
Currently, the codebase supports:
- **Patient-level assignment**: Patients are assigned to train/valid/test sets to prevent data leakage
- **Snapshot-level filtering**: `single_snapshot_per_visit=True` selects one snapshot per visit to avoid within-visit data leakage

However, there's no option to limit to **one visit per patient** within their assigned set. This would be more consistent with the existing logic, and could be useful for reducing model complexity and/or addressing non-independence when patients have many visits

## Current Behaviour
A patient assigned to the training set will contribute **all their visits** that fall within the training time period. For example:
- Patient A assigned to training set
- Has 7 visits in training period → 7 training examples (if `single_snapshot_per_visit=True`)

## Proposed Behaviour
Add `single_visit_per_patient` parameter that, when enabled, randomly selects one visit per patient within their assigned set:
- Patient A assigned to training set  
- Has 7 visits in training period → 1 randomly selected training example

### Phase 1: Core Functionality
1. **Add to `prepare_patient_snapshots()`** (primary implementation)
   ```python
   def prepare_patient_snapshots(
       df,
       prediction_time,
       exclude_columns=[],
       single_snapshot_per_visit=True,
       single_visit_per_patient=False,  # NEW
       visit_col=None,
       patient_id=None,  # NEW - required when single_visit_per_patient=True
       label_col="is_admitted",
       seed=42,
   ):
   ```

2. **Create `select_one_visit_per_patient()` helper function**
   - Follow same pattern as `select_one_snapshot_per_visit()`
   - Use random selection with seed for reproducibility
   - Group by `patient_id`, select visit with max random number

### Phase 2: Parameter Propagation
Update functions to accept and pass through the new parameter:

3. **Training functions**:
   - `train_classifier()`
   - `train_multiple_classifiers()`
   - Emergency demand training functions

4. **Visualization functions**:
   - `calibration.py` functions
   - `estimated_probabilities.py` functions  
   - `madcap.py` functions (3 calls)
   - `shap.py` functions

### Phase 3: Testing & Documentation
5. **Add comprehensive tests**
   - Unit tests for `select_one_visit_per_patient()`
   - Integration tests for `prepare_patient_snapshots()`
   - Parameter validation tests
   - End-to-end training tests

6. **Update documentation**
   - Function docstrings
   - Parameter descriptions
   - Usage examples

### Selection Logic
```python
if single_visit_per_patient and patient_id is not None:
    # After single_snapshot_per_visit filtering
    df_single = select_one_visit_per_patient(df_single, patient_id, visit_col, seed)
```

### Parameter Validation
- Require `patient_id` parameter when `single_visit_per_patient=True`
- Validate that `patient_id` column exists in DataFrame
- Maintain backward compatibility with `single_visit_per_patient=False` default

### Integration
- Works with existing `single_snapshot_per_visit` functionality
- Preserves temporal splitting integrity
- Maintains patient assignment logic from `create_temporal_splits()`

## Files to Modify
- `src/patientflow/prepare.py` (primary implementation)
- `src/patientflow/train/classifiers.py` 
- `src/patientflow/train/emergency_demand.py`
- `src/patientflow/viz/*.py` (all visualization modules)
- `tests/test_*.py` (add new tests)

## Expected Impact
- **Backward compatible**: Default behavior unchanged
- **Flexible**: Can be combined with existing filtering options
- **Consistent**: Follows established patterns in codebase
- **Reproducible**: Uses same seeding mechanism as snapshot selection

## Definition of Done
- [ ] Core functionality implemented in `prepare_patient_snapshots()`
- [ ] Helper function `select_one_visit_per_patient()` created
- [ ] All calling functions updated to support new parameter
- [ ] Comprehensive test coverage added
- [ ] Documentation updated
- [ ] Backward compatibility verified
